smithy-trait-codegen/src/it/java/software/amazon/smithy/traitcodegen/test/LoadsFromModelTest.java

```diff
@@ -92,8 +92,8 @@ static Stream<Arguments> loadsModelTests() {
                         MapUtils.of("getFieldA", Optional.of(TARGET_ONE))),
                 Arguments.of("id-ref.smithy", IdRefStructWithNestedIdsTrait.class,
                         MapUtils.of("getIdRefHolder", NestedIdRefHolder.builder().id(TARGET_ONE).build(),
-                                "getIdList", ListUtils.of(TARGET_ONE, TARGET_TWO),
-                                "getIdMap", MapUtils.of("a", TARGET_ONE, "b", TARGET_TWO))),
+                                "getIdList", Optional.of(ListUtils.of(TARGET_ONE, TARGET_TWO)),
+                                "getIdMap", Optional.of(MapUtils.of("a", TARGET_ONE, "b", TARGET_TWO)))),
```

Comment:
I'm not sure about this, I like it better returning the empty collection instead, `Optional` won't be ever empty and it will add an allocation cost per call. Probably what Jordon meant was about `@sparse` but probably he meant this. Either way, I cannot find anywhere in the source code another place where we use `Optional` with collections.

Reply:
We can know what we want when we're hand-writing implementations, but when we're generating them we should actually follow Smithy's semantics. You must be able to know if an optional member was set or not, and the only other option is to leave it null, which is not great. The way to have it generate non-optionally is to make it non-optional via `default`

Reply:
I'm talking about collections in particular, for those using `Optional` does not allow you to figure out if the member was set or not since it will be always present but empty. Testing for emptiness doesn't solve the problem either, you cannot tell apart the case when the user sets an empty list from the case where the users does not set anything at all (c.f., the java SDK, they do generate a `has<Field>()` method to find that out). Am I missing something?

Reply:
> I'm talking about collections in particular, for those using Optional does not allow you to figure out if the member was set or not since it will be always present but empty.

An empty optional was not set, an optional containing an empty list was set

Reply:
> An empty optional was not set, an optional containing an empty list was set

I'm not sure what you mean to be honest, I understand how `Optional` works. 

Allow me to elaborate what I mean, for instance this model:

```
@trait
structure CollectionsTrait {
  list: StringList
}

list StringList {
  member: String
}
```

Will currently generate a Java class as shown below, in particular the constructor will do:

```
        this.list = builder.list.copy();
```

So, for this class, the `list` field, cannot ever be null, **ever**, regardless of whether the collection was modified during build time or not. So, the following assertions will be **always** successful

1. No value was not set/added at "building" time 

```java
  assert CollectionsTrait.builder().build().getList().isPresent()
```

2. A single value was added at "building" time 

```java
  assert CollectionsTrait.builder().addList("foo").build().getList().isPresent()
```

3. The list was set to an empty collection of values

```java
  assert CollectionsTrait.builder().list(Collections.emptyList()).build().getList().isPresent()
```

So, given this, you cannot determine if the value was set by the user or not. In particular this means that this line from the getter:

```java
    public Optional<List<String>> getList() {
        return Optional.ofNullable(list);
    }
```

can be changed to:

```java
    public Optional<List<String>> getList() {
        return Optional.of(list);
    }
```


### Generated Java


```java
/**
 * Header line One
 */

package com.example.traits;

import java.util.List;
import java.util.Optional;
import software.amazon.smithy.model.node.ArrayNode;
import software.amazon.smithy.model.node.ExpectationNotMetException;
import software.amazon.smithy.model.node.Node;
import software.amazon.smithy.model.shapes.ShapeId;
import software.amazon.smithy.model.traits.AbstractTrait;
import software.amazon.smithy.model.traits.AbstractTraitBuilder;
import software.amazon.smithy.model.traits.Trait;
import software.amazon.smithy.utils.BuilderRef;
import software.amazon.smithy.utils.SmithyBuilder;
import software.amazon.smithy.utils.ToSmithyBuilder;

public final class CollectionsTrait extends AbstractTrait implements ToSmithyBuilder<CollectionsTrait> {
    public static final ShapeId ID = ShapeId.from("test.smithy.traitcodegen#CollectionsTrait");
    
    private final List<String> list;
    
    private CollectionsTrait(Builder builder) {
        super(ID, builder.getSourceLocation());
        this.list = builder.list.copy();
    }
    
    @Override
    protected Node createNode() {
        return Node.objectNodeBuilder()
            .sourceLocation(getSourceLocation()).withOptionalMember("list", getList().map(m -> m.stream().map(s -> Node.from(s)).collect(ArrayNode.collect())))
            .build();
    }
    
    /**
     * Creates a {@link CollectionsTrait} from a {@link Node}.
     * 
     * @param node Node to create the CollectionsTrait from.
     * @return Returns the created CollectionsTrait.
     * @throws ExpectationNotMetException if the given Node is invalid.
     */
    public static CollectionsTrait fromNode(Node node) {
        Builder builder = builder();
        node.expectObjectNode()
            .getArrayMember("list", n -> n.expectStringNode().getValue(), builder::list);
            
        return builder.build();
    }
    
    public Optional<List<String>> getList() {
        return Optional.ofNullable(list);
    }
    
    
    /**
     * Creates a builder used to build a {@link CollectionsTrait}.
     */
    @Override
    public SmithyBuilder<CollectionsTrait> toBuilder() {
        return builder().sourceLocation(getSourceLocation())
            .list(list);
    }
    
    public static Builder builder() {
        return new Builder();
    }
    
    /**
     * Builder for {@link CollectionsTrait}.
     */
    public static final class Builder extends AbstractTraitBuilder<CollectionsTrait, Builder> {
        private final BuilderRef<List<String>> list = BuilderRef.forList();
        
        private Builder() {}
        
        public Builder list(List<String> list) {
            clearList();
            this.list.get().addAll(list);
            return this;
        }
        
        public Builder clearList() {
            list.get().clear();
            return this;
        }
        
        public Builder addList(String value) {
            list.get().add(value);
            return this;
        }
        
        public Builder removeList(String value) {
            list.get().remove(value);
            return this;
        }
        
        @Override
        public CollectionsTrait build() {
            return new CollectionsTrait(this);
        }
    }
    
    public static final class Provider extends AbstractTrait.Provider {
        public Provider() {
            super(ID);
        }
        
        @Override
        public Trait createTrait(ShapeId target, Node value) {
            CollectionsTrait result = CollectionsTrait.fromNode(value);
            result.setNodeCache(value);
            return result;
        }
    }
}
```
```